File: 01_Language/01_Functions/php/hash_hkdf.php
<?php

// Generate a random key, and salt to strengthen it during derivation.
$inputKey = random_bytes(32);
$inputKey = "abcdef";
$salt = random_bytes(16);
$salt = "abcdef";

// Derive a pair of separate keys, using the same input created above.
$encryptionKey = hash_hkdf('sha256', $inputKey, 0, 'aes-256-encryption', $salt);
$authenticationKey = hash_hkdf('sha256', $inputKey, 0, 'sha-256-authentication', $salt);

var_dump($encryptionKey);
var_dump($authenticationKey);

var_dump($encryptionKey !== $authenticationKey); // bool(true)

File: 01_Language/01_Functions/python/hash_hkdf.py
# coding: utf-8

import hkdf
import hashlib


def hash_hkdf(algo, ikm, length=0, info=b'', salt=b''):
    h = getattr(hashlib, algo)
    kdf = hkdf.Hkdf(salt, ikm, h)
    if length == 0:
        length = h().digest_size
    return kdf.expand(info, length)


if __name__ == '__main__':
    input_key = b"abcdef"
    salt = b"abcdef"
    encryption_key = hash_hkdf("sha256", input_key, 32, b"aes-256-encryption", salt)
    authentication_key = hash_hkdf("sha256", input_key, 32, b"sha-256-authentication", salt)
    print(encryption_key)
    print(encryption_key.decode(errors="ignore"))
    print(authentication_key)
    print(authentication_key.decode(errors="ignore"))
